### ISSUE TITLE
Create DataFrame from list of rows where each row is Map

### ISSUE THREAD

**Issue by PoslavskySV** [Kotlin/dataframe#90]
Hi guys, 

it would be nice to add a method for creating a DataFrame from a list of rows represented as general `Map`s. Right now when I do:
```kotlin
val rows : List<Map<String, Any?>>

val df = rows.toDataFrame()
```
I get a wired result - DataFrame with columns obtained from the properties of `Map` class. But it is more intuitive to get a `DataFrame` with columns obtained from the keys of `Map`s. Does it make sense for you?

**Comment by koperagen** [comment]
Does this map only have primitive values? I'm asking because in that case we can simply create an overload for List<Map<String, Any?>>. But if you need to convert objects inside this Map into dataframe structures, then Map class should be supported as a special case in existing Iterable<*>.toDataFrame()

**Comment by PoslavskySV** [comment]
I am not sure that I got your question. Map values can have arbitrary type (same for each Map of iterable). For example, suppose I have a list (or any other iterable) of maps with the following structure:
```
{
    "col1" -> String,
    "col2" -> Double,
    "col3" -> MyEnum
}
```
then from the `rows.toDataFrame()` I would expect to obtain `AnyFrame` with the columns `col1` of type `String`, `col2` of type Double etc. Does it sound meaningful?

**Comment by koperagen** [comment]
Yes. Let me clarify the question. If you have col4 in your map with the type `class Name(val firstName: String, val lastName: String)`, we can convert it in two ways:
1. To `DataColumn<Name>` 
2. To `ColumnGroup`  with 2 columns, firstName and lastName (it's like `Iterable<*>.toDataFrame(depth = 2)` would work for classes)
Do you need 1 or 2?

**Comment by PoslavskySV** [comment]
Thank you for clarification, now I see! I need option 1, and generally I feel that it is more natural here.

**Comment by ian-k** [comment]
@PoslavskySV 
Till the team implements the proper solution you can use this extension function:
fun List<Map<String, Any>>.toDataFrame(): AnyFrame {
    val columns: MutableMap<String, MutableList<Any?>> = mutableMapOf()
    val notNullCols: MutableSet<String> = mutableSetOf()
    val columnSize=this.size

    this.forEachIndexed{ ri,row ->
        run {

            row.keys.forEach { col ->
                run {
                    if (columns[col] == null) {
                        columns[col] = mutableListOf()
                    }

                    val value=if (row[col].let { it is String && it.isEmpty() }) null else row[col]
                    if(value!=null)  notNullCols.add(col)

                    while(columns[col]!!.size<ri) columns[col]!!.add(null)

                    columns[col]!!.add(value)

                }
            }
        }
    }

    return columns.filter { it.value.isNotEmpty() }.filter { notNullCols.contains(it.key) }.map{ while(it.value.size<columnSize) {it.value.add(null)};it }.map { DataColumn.create(it.key ,it.value,infer = Infer.Type)  }.toDataFrame()
}

**Comment by Jolanrensen** [comment]
@ian-k I took the liberty of refactoring your example a bit so it's clearer how it works :). Thanks for sharing!

**Comment by engineerdan** [comment]
What is the work required to close this issue?  Is it just to include the [extension function](https://github.com/Kotlin/dataframe/issues/90#issuecomment-1507597926) above in the library somewhere?  Looking for a couple of 'good first issue' tickets to contribute to.

**Comment by Jolanrensen** [comment]
Since https://github.com/Kotlin/dataframe/issues/719 we now have `Map<String, *>.toDataRow()`

**Comment by Jolanrensen** [comment]
Fixed by https://github.com/Kotlin/dataframe/pull/1463

To be improved by https://github.com/Kotlin/dataframe/pull/1635
